libsolutil/JSON.cpp

```diff
@@ -22,135 +22,218 @@
+			while (j > 0 && _json[j - 1] == '\\')
+			{
+				backslashCount++;
+				j--;
+			}
+			if (backslashCount % 2 == 0)
+			{
+				inQuotes = !inQuotes;
+				fixed << c;
+				continue;
+			}
+		}
+
+		if (inQuotes)
+		{
+			if (c == '\n')
+				fixed << "\\n";
+			else if (c == '\t')
+				fixed << "\\t";
 			else
-				removeNullMembersHelper(value);
+				fixed << c;
 		}
+		else
+			fixed << c;
+	}
+	return fixed.str();
 }
 
 } // end anonymous namespace
 
-Json::Value removeNullMembers(Json::Value _json)
+Json removeNullMembers(Json _json)
 {
 	removeNullMembersHelper(_json);
 	return _json;
 }
 
-std::string jsonPrettyPrint(Json::Value const& _input)
+std::string removeNlohmannInternalErrorIdentifier(std::string const& _input)
 {
-	return jsonPrint(_input, JsonFormat{ JsonFormat::Pretty });
-}
+	std::string result = _input;
+	std::size_t startPos = result.find('[');
+	std::size_t endPos = result.find(']', startPos);
 
-std::string jsonCompactPrint(Json::Value const& _input)
-{
-	return jsonPrint(_input, JsonFormat{ JsonFormat::Compact });
+	if (startPos != std::string::npos && endPos != std::string::npos)
+		result.erase(startPos, endPos - startPos + 1);
+
+	return boost::trim_copy(result);
 }
 
-std::string jsonPrint(Json::Value const& _input, JsonFormat const& _format)
+std::string jsonPrettyPrint(Json const& _input) { return jsonPrint(_input, JsonFormat{JsonFormat::Pretty}); }
+
+std::string jsonCompactPrint(Json const& _input) { return jsonPrint(_input, JsonFormat{JsonFormat::Compact}); }
+
+std::string jsonPrint(Json const& _input, JsonFormat const& _format)
 {
-	std::map<std::string, Json::Value> settings;
-	if (_format.format == JsonFormat::Pretty)
-	{
-		settings["indentation"] = std::string(_format.indent, ' ');
-		settings["enableYAMLCompatibility"] = true;
-	}
-	else
-		settings["indentation"] = "";
-	StreamWriterBuilder writerBuilder(settings);
-	std::string result = print(_input, writerBuilder);
+	// NOTE: -1 here means no new lines (it is also the default setting)
+	std::string dumped = _input.dump(
+		/* indent */ (_format.format == JsonFormat::Pretty) ? static_cast<int>(_format.indent) : -1,
+		/* indent_char */ ' ',
+		/* ensure_ascii */ true
+	);
+
+	// let's remove this once all test-cases having the correct output.
 	if (_format.format == JsonFormat::Pretty)
-		boost::replace_all(result, " \n", "\n");
-	return result;
+		dumped = formatLikeJsoncpp(dumped, _format);
+
+	return dumped;
 }
 
-bool jsonParseStrict(std::string const& _input, Json::Value& _json, std::string* _errs /* = nullptr */)
+bool jsonParseStrict(std::string const& _input, Json& _json, std::string* _errs /* = nullptr */)
 {
-	static StrictModeCharReaderBuilder readerBuilder;
-	return parse(readerBuilder, _input, _json, _errs);
+	try
+	{
+		_json = Json::parse(
+			// TODO: remove this in the next breaking release?
+			escapeNewlinesAndTabsWithinStringLiterals(_input),
+			/* callback */ nullptr,
+			/* allow exceptions */ true,
+			/* ignore_comments */ false
```

Comment:
Turned out that we don't allow parsing comments in json files with `jsoncpp`. So I'm also disallowing this here and added a new test case `standard_import_with_comments`.

On develop this test is failing with:
```json
{
    "errors":
    [
        {
            "component": "general",
            "formattedMessage": "* Line 15, Column 14
  Syntax error: value, object or array expected.
* Line 68, Column 10
  Extra non-whitespace after JSON value.
",
            "message": "* Line 15, Column 14
  Syntax error: value, object or array expected.
* Line 68, Column 10
  Extra non-whitespace after JSON value.
",
            "severity": "error",
            "type": "JSONError"
        }
    ]
}
```

Reply:
I guess I need to add another command-line test that is only using more simple comments. The error is just referring to something from line 15 on. So maybe the first comments where ok.

Reply:
Ok, I guess we need to allow comment parsing here. Simple comments are not failing.

Reply:
However, if we allow this here - the test case is only creating an error in the jsoncpp case (more complex comments). But I guess its then ok, nlohmann seem to be able to parse these comments without creating errors.